drivers/i2c/i2c_mchp_xec.c

```diff
@@ -0,0 +1,431 @@
+{
+	const struct i2c_xec_config *config =
+		(const struct i2c_xec_config *const) (dev->config->config_info);
+	struct i2c_xec_data *data =
+		(struct i2c_xec_data *const) (dev->driver_data);
+	u32_t ba = config->base_addr;
+	u8_t byte, ctrl;
+	int ret;
+
+	/* Send slave address */
+	MCHP_I2C_SMB_DATA(ba) = (addr | BIT(0));
+	MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_ESO |
+		MCHP_I2C_SMB_CTRL_STA | MCHP_I2C_SMB_CTRL_ACK;
+	ret = _wait_completion(ba);
+	if (ret) {
+		printk("Timeout while addressing! %x\n", ret);
+		return ret;
+	}
+
+	if (msg.len == 1) {
+		/* Send NACK for last transaction */
+		MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_STA;
+	}
+
+	/* Read dummy byte */
+	byte = MCHP_I2C_SMB_DATA(ba);
+	ret = _wait_completion(ba);
+	if (ret)
+		return ret;
+
+	for (int i = 0U; i < msg.len; i++) {
+		while (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_PIN) {
+			if (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_BER) {
+				return -EBUSY;
+			}
+		}
+
+		if (i == (msg.len - 1)) {
+			if (msg.flags & I2C_MSG_STOP) {
+				/* Send stop and ack bits */
+				ctrl = (MCHP_I2C_SMB_CTRL_PIN |
+					MCHP_I2C_SMB_CTRL_ESO |
+					MCHP_I2C_SMB_CTRL_STO |
+					MCHP_I2C_SMB_CTRL_ACK);
+				MCHP_I2C_SMB_CTRL_WO(ba) = ctrl;
+				data->pending_stop = 0;
+			}
+		}
+
+		/* Send NACK for last transaction */
+		else if (i == (msg.len - 2)) {
+			MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_STA;
+		}
+		msg.buf[i] = MCHP_I2C_SMB_DATA(ba);
+	}
+
+	/* Check clock and data lines */
+	if (_check_lines(ba)) {
+		printk("Busy\n");
+		return -EBUSY;
+	}
+	return 0;
+}
+
+static int i2c_xec_transfer(struct device *dev, struct i2c_msg *msgs,
+				u8_t num_msgs, u16_t addr)
+{
+	int ret = 0;
+
+	addr <<= 1;
+
+	for (int i = 0U; i < num_msgs; i++) {
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+			ret = i2c_xec_poll_write(dev, msgs[i], addr);
+		} else {
+			ret = i2c_xec_poll_read(dev, msgs[i], addr);
+		}
+	}
+	return ret;
+}
+
+#if defined(CONFIG_I2C_SLAVE)
+static int i2c_xec_slave_register(struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static int i2c_xec_slave_unregister(struct device *dev)
+{
+	return -ENOTSUP;
+}
+#endif
+
+
+static int i2c_xec_init(struct device *dev);
+
+static const struct i2c_driver_api i2c_xec_driver_api = {
+	.configure = i2c_xec_configure,
+	.transfer = i2c_xec_transfer,
+#if defined(CONFIG_I2C_SLAVE)
+	.slave_register = i2c_xec_slave_register,
+	.slave_unregister = i2c_xec_slave_unregister,
+#endif
+};
+
+
+#ifdef CONFIG_I2C_XEC_0
+
+DEVICE_DECLARE(i2c_xec_0);
+
+static struct i2c_xec_data i2c_xec_data_0;
+static const struct i2c_xec_config i2c_xec_config_0 = {
+	.base_addr = DT_MICROCHIP_XEC_I2C_0_BASE_ADDRESS,
+	.port_sel = DT_MICROCHIP_XEC_I2C_0_PORT_SEL,
+	.irq = DT_MICROCHIP_XEC_I2C_0_DIR_IRQ,
+};
+DEVICE_AND_API_INIT(i2c_xec_0, DT_MICROCHIP_XEC_I2C_0_LABEL, &i2c_xec_init,
+			&i2c_xec_data_0, &i2c_xec_config_0,
+			POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
```

Comment:
Not sure about this one, but other i2C drivers are using the following priority 
CONFIG_KERNEL_INIT_PRIORITY_DEVICE. Can you comment @andrewboie ?

Reply:
Seems to be a KConfig I2C-specific requirement, see other I2C drivers

Reply:
@andrewboie could you comment on this priority, most of other request changes are done.

Reply:
I'm not sure either. This doesn't seem to be consistent across i2c drivers. I'm OK with leaving as-is.